### ISSUE TITLE
`jacfwd` and `jacrev` fail on different-typed inputs/outputs

### ISSUE THREAD

**Issue by romanngg** [google/jax#7780]
This example breaks both `jacfwd` and `jacrev`:

```python
def f(x):
  return x
  

x = (np.ones((), np.float16), 
     np.ones((), np.float32))
```

Then
```python
from jax import jacfwd, jacrev

jacrev(f)(x)
```
Gives
```python
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-182-404943dd2494> in <module>()
      2 
----> 3 jacrev(f)(x)

6 frames
UnfilteredStackTrace: TypeError: Type of cotangent input to vjp pullback function (float16) is not the expected tangent type (float32) of corresponding primal output with dtype float32.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
google3/third_party/py/jax/_src/api.py in _vjp_pullback_wrapper(cotangent_dtypes, cotangent_shapes, io_tree, fun, py_args)
   2047     if expected_tangent_dtype != ct_dtype:
   2048       raise TypeError(
-> 2049           f"Type of cotangent input to vjp pullback function ({ct_dtype}) is not "
   2050           f"the expected tangent type ({expected_tangent_dtype}) of corresponding primal output "
   2051           f"with dtype {_dtype(arg)}.")

TypeError: Type of cotangent input to vjp pullback function (float16) is not the expected tangent type (float32) of corresponding primal output with dtype float32.
```

And 
```python
jacfwd(f)(x)
```
Gives
```python
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-183-e245dd571352> in <module>()
----> 1 jacfwd(f)(x)

6 frames
UnfilteredStackTrace: TypeError: primal and tangent arguments to jax.jvp do not match; dtypes must be equal, or in case of int/bool primal dtype the tangent dtype must be float0.Got primal dtype float16 and so expected tangent dtype float16, but got tangent dtype float32 instead.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
google3/third_party/py/jax/_src/api.py in _jvp(fun, primals, tangents)
   1933   for p, t in safe_zip(ps_flat, ts_flat):
   1934     if core.primal_dtype_to_tangent_dtype(_dtype(p)) != _dtype(t):
-> 1935       raise TypeError("primal and tangent arguments to jax.jvp do not match; "
   1936                       "dtypes must be equal, or in case of int/bool primal dtype "
   1937                       "the tangent dtype must be float0."

TypeError: primal and tangent arguments to jax.jvp do not match; dtypes must be equal, or in case of int/bool primal dtype the tangent dtype must be float0.Got primal dtype float16 and so expected tangent dtype float16, but got tangent dtype float32 instead.
```

I believe the Jacobian should be
```python
((1 [f16 - df1/dx1], 0 [f32 - df1/dx2]),
 (0 [f16 - df2/dx1], 1 [f32 - df2/dx2]))
```

**Comment by froystig** [comment]
I wonder whether this is addressed by #7158. That PR was meant to fix #7157, which is about mixed float/complex types, but hopefully also addresses the case where distinct floating point types are mixed as well.

It is currently waiting on my closer review in response to changes made from the previous round, and doing that is on my queue... Meanwhile @NeilGirdhar has kindly been rebasing it regularly. Might be worth seeing if it fixes this too.

@NeilGirdhar, any thoughts?

**Comment by NeilGirdhar** [comment]
Yes, #7158 is supposed to address heterogenous types as well!  I added the test in this issue in case that helps.  Thanks for making the connection :smile:
